### PR TITLE
[MERGE] mail, various: improve language support in mail composer

### PR DESCRIPTION
**Pull request by tde-banana-odoo**
CONTEXT

Followup of odoo/odoo@2d791e8afeee1b60dab670b8db79be9619cbcaa6 (posting, notifying and composer / template link code
cleaning) and odoo/odoo@0b8d4c0eb1b3c0e1bb2f2de9692938015cae8ab5 (composer fields, template sync and values generation
cleaning).

RATIONALE

Naming conventions
  * content = core message / subject (+ other fields);
  * email layout = layouting applied to messages when sent to recipients who
    receive notifications by email e.g. access button, ...;
  * comment mode / mailing mode: composer main composition mode: posting on
    record(s) / sending a mailing on records;

Content situation when using a template on the composer

  * comment, monorecord mode: choose a template, content is rendered directly
    from template side according to 'lang' field definition -> ok;
  * comment, multirecord mode / mailing mode: choose a template, content is
    the raw content. At post / send, content is rendered on composer side.
    As composer is not translated -> no translation -> ko;

Email layout situation when posting / mailing

  * comment, monorecord model, using a template from UX: layout is partly
    translated based on template 'lang' field to match the content translation.
    Part of the layout still uses the current user lang;
  * other comment use cases: layout is translated based on current user;
  * mailing mode: layout not supported;

EXPECTED SITUATION

Content situation when using a template to post on a composer

  * comment, monorecord mode: choose a template, content is rendered directly
    from template side according to 'lang' field definition -> ok;
  * comment, multirecord mode / mailing mode: choose a template, content is
    the raw content. At post / send, content is rendered and translated if
    matching template content -> ok;

Email layout situation when posting / mailing

  * comment: layout translated based on recipient lang or fallback on template
    'lang' field definition, or current user's lang;
  * mailing: support layout through composer, use template 'lang' field definition
    or fallback on current user's lang;

For more details, see sub commits with their detailed explanations.

LINKS

Task-3046371 (Mail: Better Language Support in Composer)
Task-2742033 (Mail: Use recipient lang in template contextual action)
Task-2555155 (Mail: Translate notification action/access buttons)
Task-3186426 (Mail: Support notification layout in template / email composer)

### CODE REVIEW

File: addons/test_mail/tests/test_mail_composer.py [hunk 1]
```diff
@@ -1730,109 +1843,149 @@ def test_mail_composer_wtpl_complete(self):
...
+                    ctx['default_email_layout_xmlid'] = email_layout_xmlid
 
-            # global outgoing: emails not sent due to scheduled_date
-            self.assertEqual(len(self._new_mails), 2, 'Should have created 1 mail.mail per record')
-            self.assertEqual(len(self._mails), 0, 'Should not send emails, scheduled in the future')
-            self.assertEqual(self._new_mails.mapped('scheduled_date'),
-                             [self.reference_now + timedelta(days=2)] * 2)
+                # launch composer in mass mode
+                composer_form = Form(self.env['mail.compose.message'].with_context(ctx))
+                composer = composer_form.save()
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:1907]: Don't really see the point of using all these defaults and a "Form", can't we just create a record?

Or do we need to trigger some onchange or something to make it work?

File: addons/test_mail/tests/test_mail_composer.py [hunk 2]
```diff
@@ -1730,109 +1843,149 @@ def test_mail_composer_wtpl_complete(self):
...
 
-            # simulate cron queue at right time for sending
-            with freeze_time(self.reference_now + timedelta(days=2)):
-                self.env['mail.mail'].sudo().process_email_queue()
+                # ensure some parameters used afterwards
+                author = self.env.user.partner_id
+                self.assertEqual(composer.author_id, author,
+                                 'Author cannot be synchronized with a raw email_from')
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:1912]: (Strange explanation)

File: addons/test_mail/tests/test_mail_composer.py [hunk 3]
```diff
@@ -1730,109 +1843,149 @@ def test_mail_composer_wtpl_complete(self):
...
 
-        # hack to use assertEmails: filtering on from/to only is not sufficient to distinguish emails
-        _mails_records = [
-            [mail for mail in self._mails if '%s-%s' % (record.id, record._name) in mail['message_id']]
-            for record in self.test_records
-        ]
+                    new_partners = self.env['res.partner'].search([
+                        ('email', 'in', [email_to_1, email_to_2, email_to_3, email_cc_1])
+                    ])
+                    self.assertEqual(len(new_partners), 3)
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:1922]: (Not testing their lang this time?)

File: addons/test_mail/tests/test_mail_composer.py [hunk 4]
```diff
@@ -1730,109 +1843,149 @@ def test_mail_composer_wtpl_complete(self):
...
+
+                    if not scheduled_date:
+                        # global outgoing: emails sent
+                        self.assertEqual(len(self._new_mails), 2, 'Should have created 1 mail.mail per record')
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:1926]: Could move this out of the if then.

File: addons/test_mail/tests/test_mail_composer.py [hunk 5]
```diff
@@ -1730,109 +1843,149 @@ def test_mail_composer_wtpl_complete(self):
...
+                                         [self.reference_now + timedelta(days=2)] * 2)
+
+                        # simulate cron queue at right time for sending
+                        with freeze_time(self.reference_now + timedelta(days=2)):
+                            self.env['mail.mail'].sudo().process_email_queue()
 
-        for record, _mails in zip(self.test_records, _mails_records):
-            # message copy is kept
-            message = record.message_ids[0]
+                self.assertEqual(len(self._mails), 10, 'Should have sent 5 emails per record')
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:1941]: Was already tested for the "not scheduled_date" case so could be moved inside the else part ?

File: addons/test_mail/tests/test_mail_composer.py [hunk 6]
```diff
@@ -1730,109 +1843,149 @@ def test_mail_composer_wtpl_complete(self):
...
-                                  default_template_id=self.template.id)
-        ))
-        composer = composer_form.save()
-        with self.mock_mail_gateway(mail_unlink_sent=True):
-            # remove alias so that _notify_get_reply_to will return the default value instead of alias
-            self.env['ir.config_parameter'].sudo().set_param("mail.catchall.domain", None)
-            composer.action_send_mail()
+                    # translation are currently not supported at all as they
+                    # are fetched composer side, which is most probably not
+                    # translated (only template is)
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py]: "most probably" ?

File: addons/test_mail/tests/test_mail_composer.py [hunk 7]
```diff
@@ -1730,109 +1843,149 @@ def test_mail_composer_wtpl_complete(self):
...
+                    if False and exp_lang == 'es_ES':
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py]: Strange but I guess it's to be able to enable it easily later.

File: addons/test_mail/tests/test_message_post.py [hunk 8]
```diff
@@ -237,25 +238,14 @@ def test_notify_recipients_internals(self):
...
-            emp_info = next(item for item in classify_res if item['recipients'] == self.partner_employee.ids)
-            # check there is no access button
-            self.assertFalse(partner_info['has_button_access'])
-            self.assertFalse(emp_info['has_button_access'])
-
-            # test on falsy records (False model cannot be browsed, skipped)
-            if model:
-                record_falsy = self.env[model].browse(res_id)
-                classify_res = record_falsy._notify_get_recipients_classify(
+            with self.subTest(model=model, res_id=res_id):
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_message_post.py:244]: Missed a zip opportunity here :disguised_face:

File: addons/test_mail/tests/test_message_post.py [hunk 9]
```diff
@@ -264,6 +254,18 @@ def test_notify_recipients_internals(self):
...
+                if model:
+                    record_falsy = self.env[model].browse(res_id)
+                    classify_res = record_falsy._notify_get_recipients_classify(
+                        pdata, 'Test', msg_vals=notify_msg_vals)
+                    # find back information for partner
+                    partner_info = next(item for item in classify_res if item['recipients'] == self.partner_1.ids)
+                    emp_info = next(item for item in classify_res if item['recipients'] == self.partner_employee.ids)
+                    # check there is no access button
+                    self.assertFalse(partner_info['has_button_access'])
+                    self.assertFalse(emp_info['has_button_access'])
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_message_post.py:274]: Seems duplicated, but maybe this test will be modified again in a later commit to introduce some changes based on model / no model.

File: addons/test_mail/tests/test_message_post.py [hunk 10]
```diff
@@ -1477,46 +1481,103 @@ def setUpClass(cls):
...
-        self.assertIn('View Lang Chatter Model', body,
-                      'Fixme: "View document" should be translated')
-        # self.assertIn('SpanishView Spanish Model Description', body,
-        #               '"View document" should be translated')
-        # self.assertNotIn(f'View {test_record._description}', body,
-        #                  '"View document" should be translated')
-        # self.assertIn('SpanishButtonTitle', body,
-        #               'Groups-based action names should be translated')
-        self.assertIn('NotificationButtonTitle', body, 'Fixme: Groups-based action names should be translated')
+        for partner, exp_content_lang, exp_notif_lang in zip(
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_message_post.py:1516]: Don't see why it's "exp_" here, it's the test variables, not the expected result.

File: addons/test_mail/tests/test_message_post.py [hunk 11]
```diff
@@ -1477,46 +1481,103 @@ def setUpClass(cls):
...
+            (self.env['res.partner'], self.partner_1, self.partner_2),
+            ('es_ES', 'en_US', 'es_ES'),  # content: coming from template definition
+            ('en_US', 'en_US', 'es_ES'),  # layout: coming from customer
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_message_post.py:1519]: Actually not sure why those are here.

It's not given to the "message_post_with_source" call so it's the values of something on the template / customer?
Should we assert that against the related fields then? (Note that it seems done for the customer but not for the template definition)
Or can't we use the related fields instead of giving this as test input?
There is only 1 template yet the values are changing multiple times?
It's a bit confusing.

File: addons/test_mail/tests/test_message_post.py [hunk 12]
```diff
@@ -1624,3 +1691,93 @@ def test_layout_email_lang_template(self):
...
+                                'email_from': self.partner_employee.email_formatted,
+                                'subject': 'TeDeum',
+                            },
+                           )
+
+                        # Low-level checks on outgoing email for the recipient to
+                        # check layouting and language. Note that standard layout
+                        # is not tested against translations, only the custom one
+                        # to ease translations checks.
+                        for partner in self.partner_1 + self.partner_2:
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_message_post.py]: (Is there a point to sending them to multiple partners by the way? Maybe it comes in a later commit)

File: addons/test_mail/tests/test_message_post.py [hunk 13]
```diff
@@ -1624,3 +1691,93 @@ def test_layout_email_lang_template(self):
...
+                                    self.assertIn(exp_action_en, email['body'])
+                            else:
+                                # check default layouting applies
+                                if employee_lang == 'es_ES':
+                                    self.assertIn('html lang="es_ES"', email['body'])
+                                elif employee_lang:
+                                    self.assertIn('html lang="en_US"', email['body'])
+                                else:
+                                    # if lang is False -> nothing in header, strange
+                                    self.assertNotIn('html lang', email['body'])
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_message_post.py]: Looking at input, not sure how we get here.

**Review by awa-odoo — COMMENTED** [review]
Bloups :fish: 

Already posting a few comments on first commit (the big add tests one).

Cheers!

File: addons/mail/models/mail_thread.py [hunk 14]
```diff
@@ -3503,7 +3583,7 @@ def _encode_link(self, base_link, params):
     def _get_model_description(self, model_name, lang):
         if not model_name:
             return False
-        IrModel = self.env['ir.model'].with_context(lang=lang or self.env.lang)
+        IrModel = self.env['ir.model'].with_context(lang=(lang or self.env.lang))
```
Comment (awa-odoo) [inline addons/mail/models/mail_thread.py:3586]: Should that be in an earlier commit?

File: addons/test_mail/tests/test_mail_composer.py [hunk 15]
```diff
@@ -1490,10 +1504,15 @@ def test_mail_composer_wtpl_complete(self):
...
 
-                        # global outgoing: one mail.mail (all customer recipients, then all employee recipients)
-                        # and 5 emails, and 1 inbox notification (admin)
-                        self.assertEqual(len(self._new_mails), 2 * len(test_records),
-                                         'Should have created 1 mail.mail per group per record')
+                        # global outgoing: one mail.mail (all customer recipients,
+                        # then all employee recipients, maybe split / lang) and 5
+                        # emails, and 1 inbox notification (admin)
+                        if use_lang:
+                            self.assertEqual(len(self._new_mails), 2 * len(test_records) + 1,
```
Comment (awa-odoo) [inline addons/test_mail/tests/test_mail_composer.py:1511]: Would be nice to re-explain here why "+1", a bit hard to follow.

File: addons/mail/wizard/mail_compose_message.py [hunk 16]
```diff
@@ -992,6 +992,45 @@ def _prepare_mail_values_dynamic(self, res_ids):
...
+                    model_description=False,  # force dynamic computation
+                    force_email_lang=lang,
+                ):
+                    mail_body = record_wlang._notify_by_email_render_layout(
+                        message_inmem,
+                        recipients_group_data,
+                        msg_vals=msg_vals,
+                        render_values=render_values,
+                    )
+                    mail_values['body_html'] = mail_body
```
Comment (awa-odoo) [inline addons/mail/wizard/mail_compose_message.py:1032]: Does this mean that the body is rendered once through "render_field" and then we override it here? Or do we just encapsulate the body into a proper template with the "message_inmem" thing?

Does that mean that it was not encapsulated before?
So we add multiple features here actually?

Maybe commit message is a bit light then.

File: addons/mail/wizard/mail_compose_message.py [hunk 17]
```diff
@@ -992,6 +992,45 @@ def _prepare_mail_values_dynamic(self, res_ids):
                     reply_to = mail_values.get('email_from', False)
                 mail_values['reply_to'] = reply_to
 
+            # body: render layout in email mode (comment mode is managed by the
+            # notification process, see @_notify_thread_by_email)
+            if email_mode and self.email_layout_xmlid and mail_values['recipient_ids']:
+                lang = langs[res_id]
+                recipient_ids = [command[1] for command in mail_values['recipient_ids']]
```
Comment (awa-odoo) [inline addons/mail/wizard/mail_compose_message.py:999]: So we expect only 4, ID commands and not anything else?

**Review by tde-banana-odoo — COMMENTED** [review]
@awa-odoo answers / udpated "big test commit" comments, but on https://github.com/odoo/odoo/pull/114511 (which is the forward port of tests and bugfixes, I included a smaller versions of tests in stable and the complete one in the forward port directly).

File: addons/mail/wizard/mail_compose_message.py [hunk 18]
```diff
@@ -992,6 +992,45 @@ def _prepare_mail_values_dynamic(self, res_ids):
...
+                message_inmem = self.env['mail.message'].new({
+                    'body': mail_values['body'],
+                })
+
+                for _lang, render_values, recipients_group_data in record_wlang._notify_by_email_iterate_recipients(
+                    message_inmem,
+                    [{
+                        'active': True,
+                        'id': pid,
+                        'lang': False,
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/wizard/mail_compose_message.py]: Indeed, I think we can force the lang here.

[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/wizard/mail_compose_message.py]: Note that the record.with_context also allows to translate the record itself, if for example we have something like '<span t-out="object.partner_id.title"', we have to have a context on the record itself when rendering.

File: addons/mail/wizard/mail_compose_message.py [hunk 19]
```diff
@@ -992,6 +992,45 @@ def _prepare_mail_values_dynamic(self, res_ids):
...
+            # notification process, see @_notify_thread_by_email)
+            if email_mode and self.email_layout_xmlid and mail_values['recipient_ids']:
+                lang = langs[res_id]
+                recipient_ids = [command[1] for command in mail_values['recipient_ids']]
+                record_wlang = record.with_context(lang=lang)
+
+                msg_vals = {
+                    'email_layout_xmlid': self.email_layout_xmlid,
+                    'model': record_wlang._name,
+                    'res_id': record_wlang.id,
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/wizard/mail_compose_message.py]: We can at least use the composer.model / res_id values for model and res_id.

File: addons/mail/wizard/mail_compose_message.py (hunk 16 above)
Comment (tde-banana-odoo) [reply addons/mail/wizard/mail_compose_message.py:1027]: Indeed it is a complete new feature. I forgot to update the commit message, but currently if you use the composer in mass mailing mode, it just sends the body. If you use the composer to post a message, emails are sent with a layout. Here we now support using a layout in mailing mode, to try to make both modes more coherent.

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, answered / updated most comments :)

File: addons/test_mail/tests/test_mail_composer_mixin.py [hunk 20]
```diff
@@ -79,24 +79,17 @@ def test_content_sync(self):
 
         # template with void values: should not force void (TODO)
         composer.template_id = template_void.id
-        self.assertEqual(composer.body, f'<p><br /></p>',
-                         'TODO: should not force void value')
+        self.assertEqual(composer.body, '<p>CustomBody for <t t-out="object.name"/></p>')
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/test_mail/tests/test_mail_composer_mixin.py:82]: Ok, my point was that if you select template A and then select template B, if they have some values for some fields and not for others, it seems strange to mix them.

To me it should be the same use case as removing, as you end-up with mixed composer values from A and B (which is close to "half backed" as mentioned when you set to void).

But that's a niche thing, and maybe a matter of taste.

File: addons/mail/wizard/mail_compose_message.py [hunk 21]
```diff
@@ -744,6 +757,7 @@ def _prepare_mail_values(self, res_ids):
             DYN - 'body',
             STA - 'email_add_signature',
             STA - 'email_layout_xmlid',
+            DYN - 'force_email_lang',  # notify parameter
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/mail/wizard/mail_compose_message.py:759]: I thought dynamic was actually evaluated, like a qweb expression, hence the confusion.
Still not super clear to me but maybe it's a general feeling, not particularly related to this param addition, so it's fine if you want to leave it like this.

File: addons/mail/models/mail_thread.py [hunk 22]
```diff
@@ -2908,37 +2908,9 @@ def _notify_thread_by_email(self, message, recipients_data, msg_vals=False,
         if not partners_data:
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/mail/models/mail_thread.py:3012]: Did not get an answer to this yet I feel like the questions are important ^^

I guess I could take an example to be more specific:

Considering a user in English.
That uses the composer to send a notification on a Lead record, which langage is French.
With Follower A and Follower B, that have respectively German and Spanish as language.
The template used in the composer is fully translated in every language and not modified before sending.

In what language will the received notification be for the followers?
Is it going to be the same language for the layout part (buttons etc) and for the content part?

Sorry if this is explicitly covered by tests, a bit hard to follow so I prefer a confirmation here.

EDIT: maybe you were still working on this part actually, anyway example will help me understand ^^

File: addons/mail/models/mail_thread.py [hunk 23]
```diff
@@ -3114,8 +3111,90 @@ def send_notifications():
 
         return True
 
-    def _notify_by_email_prepare_rendering_context(self, message, msg_vals=False, model_description=False,
-                                                   force_email_company=False, force_email_lang=False):
+    def _notify_get_recipients_and_context_iterator(
```
Comment (awa-odoo) [inline addons/mail/models/mail_thread.py:3114]: _notify_get_classified_recipients_iterator?

"and_context" is a bit strange, I feel like adding "classified" respects other naming more.

File: addons/mail/models/mail_thread.py [hunk 24]
```diff
@@ -3029,6 +3007,69 @@ def send_notifications():
...
+            render_values: used to render the notification layout and translated
+            using lang,
+            recipients_group_data: a recipients group which is a dict containing
+            data defined in "_notify_get_recipients",
+          );
+        """
+        lang_to_recipients = {}
+        for data in recipients_data:
+            lang_to_recipients.setdefault(
+                data.get('lang') or force_email_lang or self.env.lang,
```
[the remark this answers is not in the document]
Comment (awa-odoo) [reply addons/mail/models/mail_thread.py:3164]: That's fair ^^

File: addons/mail/models/mail_thread.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py:3012]: Did not see the message actually ^^ already partially answered, but to be clear (and I will add an example in merge message): we now translate the layout in the recipient lang, not current user.Rationale is: this is the context sent by Odoo, so it should be in the same lang as your user (aka french if  you speak french), whatever the content language.

In short, previously, Luc was receiving Thomas messages with a french layout (Thomas language). A bit weird, as he may not understand it. Now he receives it in English (its own language).

When the content language is forced through template lang field, it is used as default value if the recipient has no lang (aka: template translated in spanish for recipients not having a lang -> spanish layout to try to be coherent).

So, for your example: content will be translated based on template definition (generally partner_id lang). Then the layout will be translated into German and Spanish for the followers (if they receive emails obviously). So they receive the content in spanish (targets customer) but the layout in their chosen language (to understand the context).

It is effectively covered by tests, the tests where everything was not clear when reading it :D .

Is it clearer ?

File: addons/test_mail/tests/test_mail_composer_mixin.py (hunk 20 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer_mixin.py:82]: Good point. Our hypothesis is that a void template never overrides user input. Here we cannot really know the content comes from an old template, you may have typed it by hand (and don't want to try with _origin.template_id.body, would make code even more complicated). So if you change template during editing, indeed you may have half baked content... but hey. We did what we could :D .

[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/test_mail/tests/test_mail_composer_mixin.py:82]: Added a comment in commit message.

File: addons/mail/models/mail_render_mixin.py [hunk 25]
```diff
@@ -203,6 +203,11 @@ def _prepend_preview(self, html, preview):
             return tools.prepend_html_content(html, html_preview)
         return html
 
+    def _valid_render_options(self):
+        """ Set of valid rendering options. To be sure it is called with the
+        right arguments. Can be overridden to add specific support. """
+        return {'post_process', 'preserve_comments'}
```
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/models/mail_render_mixin.py]: In the end it is not overriden anymore, back to fixed arguments.

File: addons/mail/wizard/mail_compose_message.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/wizard/mail_compose_message.py:759]: In code, I explained a bit difference between 'dynamic' (depends on res_ids) and 'rendered' (depends on qweb). You may be dynamic without being rendered :) . Not saying all code is crystal clear, but that's the wording we are going to use I think (probably some docstrings yet to improve).

File: addons/mail/models/mail_thread.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py:3012]: Added a rationale + naming + example in commit

File: addons/mail/models/mail_thread.py (hunk 23 above)
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py]: Ok.

**Review by tde-banana-odoo — COMMENTED** [review]
Answers-that-are-not-spammed-because-I-found-the-original-discussion

**Comment by tde-banana-odoo** [comment]
@robodoo override=ci/security

Reason: just moving code, no change.

**Comment by tde-banana-odoo** [comment]
@robodoo r+ rebase-merge
